raft: step down immediately on leader transfer

nvb opened the pull request:
Part of #129807.

This commit updates the handling of raft leadership transfers to have the outgoing leader step down immediately after the MsgTimeoutNow is sent. We retain the `leadTransferee` field because there is still a "pending", reversible transfer state while the leader is catching the follower up on its log. However, once the catch up is complete, the leader now steps down, as it has irrevocably compromised its leadership term by giving the target permission to overthrow it.

If the transfer fails after the MsgTimeoutNow has been sent, the leader (who has stepped down to a follower) must call a new election at a new term in order to reestablish leadership.

Release note: None

nvb committed:
raft: clean up early termination in tickHeartbeat on leader step down

Instead of checking for a leader step down multiple times in the middle of
tickHeartbeat, check once on the code path that could cause the leader to
step down (check quorum). This logic dates back to 65b78dc1 and 3d04e279.

Epic: None
Release note: None

nvb committed:
raft: assert tick functions are called with expected states

This commit adds assertions on `raft.state` at the beginning of
`tickElection` and `tickHeartbeat`.

Epic: None
Release note: None

nvb committed:
raft: step down immediately on leader transfer

Part of #129807.

This commit updates the handling of raft leadership transfers to have the
outgoing leader step down immediately after the MsgTimeoutNow is sent. We
retain the `leadTransferee` field because there is still a "pending",
reversible transfer state while the leader is catching the follower up on
its log. However, once the catch up is complete, the leader now steps down,
as it has irrevocably compromised its leadership term by giving the target
permission to overthrow it.

If the transfer fails after the MsgTimeoutNow has been sent, the leader
(who has stepped down to a follower) must call a new election at a new
term in order to reestablish leadership.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/130123)
<!-- Reviewable:end -->

nvb committed:
raft: test against MsgTimeoutNow from old terms

This commit adds a new `TestLeaderTransferDifferentTerms`, which verifies
that a MsgTimeoutNow will only be respected if it is from the current term
or from a new term.

Epic: None
Release note: None

arulajmani reviewed (commented):
Mostly looks good. Talking about this in the pod meeting yesterday and then coming to review the PR made it a breeze

> Reviewed 1 of 1 files at r1, 1 of 1 files at r2, 3 of 3 files at r3, 2 of 2 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/130123#-:-O66Z6E-CoWD9A3YMpf4:b-z09iu6)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten)

___
*[`pkg/raft/raft.go` line 372 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/130123#-O66LKPR3I45B53E7qCQ:-O66LKPR3I45B53E7qCR:b-3acy2k) ([raw file](https://github.com/cockroachdb/cockroach/blob/697741d718f878719930cf2179b4732eead9753e/pkg/raft/raft.go#L372)):*
> ```Go
> 	// leadTransferee, if set, is the id of the leader transfer target during a
> 	// pending leadership transfer. The value is set while the outgoing leader
> 	// (this node) is catching the target up on its log. Once the target is caught
> ```

Consider adding some words about how this field is used when the leader is catching up the target. You mentioned in yesterday's pod meeting that it's to reject incoming proposals, right? 

nit: maybe a [1] pattern that we use elsewhere to not distract from the main idea of the comment.
___
*[`pkg/raft/raft.go` line 376 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/130123#-O66MMKy1VnGj1YrIpkn:-O66MMKy1VnGj1YrIpko:bv1u9yr) ([raw file](https://github.com/cockroachdb/cockroach/blob/697741d718f878719930cf2179b4732eead9753e/pkg/raft/raft.go#L376)):*
>	// it has irrevocably compromised its leadership term by giving the target
	// permission to overthrow it.

ChatGPT couldn't have dramatized this better :joy: 
___
*[`pkg/raft/raft.go` line 1760 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/130123#-O66N4WYDY_Z9QprF1YD:-O66N4WYDY_Z9QprF1YE:bgivcah) ([raw file](https://github.com/cockroachdb/cockroach/blob/697741d718f878719930cf2179b4732eead9753e/pkg/raft/raft.go#L1760)):*
> ```Go
> 		// Transfer leadership to third party.
> 		r.logger.Infof("%x [term %d] starts to transfer leadership to %x", r.id, r.Term, leadTransferee)
> 		// Transfer leadership should be finished in one electionTimeout, so reset r.electionElapsed.
> ```

nit: should we expand this comment to say that if it doesn't, the outgoing leader is free to campaign at a higher term.
___
*[`pkg/raft/raft.go` line 1768 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/130123#-O66Ne9Q3e1OPNmOf0AJ:-O66Ne9Q3e1OPNmOf0AK:b-l0er6u) ([raw file](https://github.com/cockroachdb/cockroach/blob/697741d718f878719930cf2179b4732eead9753e/pkg/raft/raft.go#L1768)):*
> ```Go
> 		} else {
> 			pr.MsgAppProbesPaused = false
> 			r.maybeSendAppend(leadTransferee)
> ```

Consider adding a comment here that explains how we handle the state transition when we're trying to transfer leadership to a follower that's behind on it's log. Might be worth including a pointer to `MsgAppResp` where we call `transferLeader`? 

All of this was very easy to follow along because we talked about this yesterday and you've got a great explanation in your commit message, but if I were just reading the code, I'd think this was a bit under documented and subtle.
___
*[`pkg/raft/raft.go` line 2301 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/130123#-O66P5k7BxG1zpFs72oO:-O66P5k7BxG1zpFs72oP:b5lamqp) ([raw file](https://github.com/cockroachdb/cockroach/blob/697741d718f878719930cf2179b4732eead9753e/pkg/raft/raft.go#L2301)):*
> ```Go
> }
> 
> func (r *raft) transferLeader(to pb.PeerID) {
> ```

Should we add an assertion to say `r.state == StateLeader && r.lead == r.id`?
___
*[`pkg/raft/raft.go` line 2306 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/130123#-O66Woel1YsILsTwJzho:-O66Woel1YsILsTwJzhp:bw4nk40) ([raw file](https://github.com/cockroachdb/cockroach/blob/697741d718f878719930cf2179b4732eead9753e/pkg/raft/raft.go#L2306)):*
> ```Go
> }
> 
> func (r *raft) abortLeaderTransfer() {
> ```

Should we add the assertion from above here as well? 
___
*[`pkg/raft/testdata/confchange_v2_replace_leader.txt` line 158 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/130123#-O66WWJ42AVY2Aul8wdG:-O66WWJ42AVY2Aul8wdH:bam0ypa) ([raw file](https://github.com/cockroachdb/cockroach/blob/697741d718f878719930cf2179b4732eead9753e/pkg/raft/testdata/confchange_v2_replace_leader.txt#L158)):*
> ```help
> raft-state
> ----
> 1: StateFollower (Voter) Term:1 Lead:0
> ```

Unfortunate how the `Lead: 0` diff came about due to https://github.com/cockroachdb/cockroach/blob/c921401097652a5864318ab8a6c068b451586d84/pkg/raft/status.go#L103-L119
___
*[`pkg/raft/raft_test.go` line 3201 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/130123#-O66YmaY5NkJf7z0IrC9:-O66YmaZ3lPRMGptRlK1:b-i4sl8g) ([raw file](https://github.com/cockroachdb/cockroach/blob/b7fe3944f5dec6cc945727be6f3192df5fe59d97/pkg/raft/raft_test.go#L3201)):*
> ```Go
> 		}
> 		require.Equal(t, expState, r.state)
> 		require.Equal(t, uint64(3), r.Term)
> ```

Do you think it's worth adding a variant of this test where the node that receives the `MsgTimeoutNow` (n1) from `n2, T2` hasn't heard about n3 being leader at `T3`? 


<!-- Sent from Reviewable.io -->

nvb committed:
raft: test against delayed MsgTimeoutNow on stale follower

This commit adds a new `TestLeaderTransferStaleFollower`, which verifies
that a MsgTimeoutNow received by a stale follower (a follower still at
an earlier term) will cause the follower to call an election which it
can not win.

Epic: None
Release note: None

nvb reviewed (commented):
TFTR!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/130123#-:-O67VzFqAnm4e3_C6o2A:b-814g8k)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani)

___
*[`pkg/raft/raft.go` line 372 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/130123#-O66LKPR3I45B53E7qCQ:-O67L7b07x6hPx-tNWVt:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/697741d718f878719930cf2179b4732eead9753e/pkg/raft/raft.go#L372)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Consider adding some words about how this field is used when the leader is catching up the target. You mentioned in yesterday's pod meeting that it's to reject incoming proposals, right? 

nit: maybe a [1] pattern that we use elsewhere to not distract from the main idea of the comment.
</blockquote></details>

Done.
___
*[`pkg/raft/raft.go` line 376 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/130123#-O66MMKy1VnGj1YrIpkn:-O67L7soC0piiQXuvJyt:b-d23daf) ([raw file](https://github.com/cockroachdb/cockroach/blob/697741d718f878719930cf2179b4732eead9753e/pkg/raft/raft.go#L376)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

>	// it has irrevocably compromised its leadership term by giving the target
	// permission to overthrow it.

ChatGPT couldn't have dramatized this better :joy:
</blockquote></details>

There's no point writing code comments if you aren't going to make them dramatic 😂
___
*[`pkg/raft/raft.go` line 1760 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/130123#-O66N4WYDY_Z9QprF1YD:-O67Lh2E75sYaIRQPKeq:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/697741d718f878719930cf2179b4732eead9753e/pkg/raft/raft.go#L1760)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

nit: should we expand this comment to say that if it doesn't, the outgoing leader is free to campaign at a higher term.
</blockquote></details>

Done.
___
*[`pkg/raft/raft.go` line 1768 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/130123#-O66Ne9Q3e1OPNmOf0AJ:-O67MVrM-A-EiUmbCk_c:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/697741d718f878719930cf2179b4732eead9753e/pkg/raft/raft.go#L1768)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Consider adding a comment here that explains how we handle the state transition when we're trying to transfer leadership to a follower that's behind on it's log. Might be worth including a pointer to `MsgAppResp` where we call `transferLeader`? 

All of this was very easy to follow along because we talked about this yesterday and you've got a great explanation in your commit message, but if I were just reading the code, I'd think this was a bit under documented and subtle.
</blockquote></details>

Done.
___
*[`pkg/raft/raft.go` line 2301 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/130123#-O66P5k7BxG1zpFs72oO:-O67NGQhDIh6BXPpGHs7:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/697741d718f878719930cf2179b4732eead9753e/pkg/raft/raft.go#L2301)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Should we add an assertion to say `r.state == StateLeader && r.lead == r.id`?
</blockquote></details>

Done.
___
*[`pkg/raft/raft.go` line 2306 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/130123#-O66Woel1YsILsTwJzho:-O67NNFI5pmLDpJKNgjB:b7cvh3e) ([raw file](https://github.com/cockroachdb/cockroach/blob/697741d718f878719930cf2179b4732eead9753e/pkg/raft/raft.go#L2306)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Should we add the assertion from above here as well?
</blockquote></details>

We call this in `reset` as well, so it doesn't necessarily assume that a leadership transfer is in progress. I'll leave as is for now.
___
*[`pkg/raft/raft_test.go` line 3201 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/130123#-O66YmaY5NkJf7z0IrC9:-O67VaPy9bzAH_jLigzQ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b7fe3944f5dec6cc945727be6f3192df5fe59d97/pkg/raft/raft_test.go#L3201)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Do you think it's worth adding a variant of this test where the node that receives the `MsgTimeoutNow` (n1) from `n2, T2` hasn't heard about n3 being leader at `T3`?
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

arulajmani reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 3 of 3 files at r5, 3 of 3 files at r6, 2 of 2 files at r7, 1 of 1 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/130123#-:-O67dEGeEjYeT5YMvDm7:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @nvanbenschoten)

___
*[`pkg/raft/raft.go` line 2306 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/130123#-O66Woel1YsILsTwJzho:-O67dFRi6FIuFQ6LtnsD:bdkxifj) ([raw file](https://github.com/cockroachdb/cockroach/blob/697741d718f878719930cf2179b4732eead9753e/pkg/raft/raft.go#L2306)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

We call this in `reset` as well, so it doesn't necessarily assume that a leadership transfer is in progress. I'll leave as is for now.
</blockquote></details>

Want to add a TODO to add this assertion in the future? Feel free to put my name on it.
___
*[`pkg/raft/raft.go` line 1773 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/130123#-O67c0GR8dZAvVQODjA-:-O67c0GR8dZAvVQODjA0:b-o6l84a) ([raw file](https://github.com/cockroachdb/cockroach/blob/9a7439725ee0e2d3957c668abfff4653aa099b16/pkg/raft/raft.go#L1773)):*
> ```Go
> 		} else {
> 			// If the transfer target is not initially caught up on its log, we don't
> 			// sent it a MsgTimeoutNow immediately. Instead, we eagerly try to catch
> ```

nit: s/sent/send/


<!-- Sent from Reviewable.io -->

nvb reviewed (commented):
bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/130123#-:-O67is4RCMWUgI0ptL2r:b-ydi9ay)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @arulajmani)

___
*[`pkg/raft/raft.go` line 2306 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/130123#-O66Woel1YsILsTwJzho:-O67iKle4FBZluxZ07wM:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/697741d718f878719930cf2179b4732eead9753e/pkg/raft/raft.go#L2306)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Want to add a TODO to add this assertion in the future? Feel free to put my name on it.
</blockquote></details>

Done.
___
*[`pkg/raft/raft.go` line 1773 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/130123#-O67c0GR8dZAvVQODjA-:-O67iK-K2Sp6dqjZtWsS:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9a7439725ee0e2d3957c668abfff4653aa099b16/pkg/raft/raft.go#L1773)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

nit: s/sent/send/
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->
